### ISSUE TITLE
heap-use-after-free in test_ssl_old

### ISSUE THREAD

**Issue by juergenchrist** [openssl/openssl#15284]
When running test_ssl_old on f04bb0bce490de847ed0482b8ec9eabedd173852 in a sanitizer build, we see this error multiple times:
```
==443426==ERROR: AddressSanitizer: heap-use-after-free on address 0x6080000000f8 at pc 0x03ff851f42b2 bp 0x03ffe167d2c0 sp 0x03ffe167d2b8
READ of size 8 at 0x6080000000f8 thread T0
    #0 0x3ff851f42b1 in ossl_provider_free_parent crypto/provider_child.c:329
    #1 0x3ff851fc6c5 in ossl_provider_free crypto/provider_core.c:478
    #2 0x3ff851fc6c5 in ossl_provider_free crypto/provider_core.c:424
    #3 0x3ff85156a77 in evp_rand_free crypto/evp/evp_rand.c:78
    #4 0x3ff85156a77 in EVP_RAND_free crypto/evp/evp_rand.c:293
    #5 0x3ff851570e7 in EVP_RAND_CTX_free crypto/evp/evp_rand.c:392
    #6 0x3ff852c6f37 in rand_ossl_ctx_free crypto/rand/rand_lib.c:476
    #7 0x3ff852c6f37 in rand_ossl_ctx_free crypto/rand/rand_lib.c:466
    #8 0x3ff851d1651 in CRYPTO_free_ex_data crypto/ex_data.c:402
    #9 0x3ff851c71c5 in context_deinit crypto/context.c:132
    #10 0x3ff851c71c5 in context_deinit crypto/context.c:115
    #11 0x3ff851c71c5 in OSSL_LIB_CTX_free crypto/context.c:243
    #12 0x3ff851c71c5 in OSSL_LIB_CTX_free crypto/context.c:238
    #13 0x3ff8644bc4f in legacy_teardown providers/legacyprov.c:161
    #14 0x135896b in ossl_provider_free crypto/provider_core.c:441
    #15 0x135896b in ossl_provider_free crypto/provider_core.c:424
    #16 0x13d3441 in OPENSSL_sk_pop_free crypto/stack/stack.c:406
    #17 0x1353315 in sk_OSSL_PROVIDER_pop_free crypto/provider_core.c:118
    #18 0x1353315 in provider_store_free crypto/provider_core.c:173
    #19 0x1353315 in provider_store_free crypto/provider_core.c:165
    #20 0x1334721 in CRYPTO_free_ex_data crypto/ex_data.c:402
    #21 0x132bb95 in context_deinit crypto/context.c:132
    #22 0x132bb95 in context_deinit crypto/context.c:115
    #23 0x132bb95 in OSSL_LIB_CTX_free crypto/context.c:243
    #24 0x132bb95 in OSSL_LIB_CTX_free crypto/context.c:238
    #25 0x1011309 in main test/ssl_old_test.c:1867
    #26 0x3ff895abb49 in __libc_start_main (/lib64/libc.so.6+0x2bb49)
    #27 0x10224fd  (/root/code/openssl/test/ssl_old_test+0x10224fd)
0x6080000000f8 is located 88 bytes inside of 96-byte region [0x6080000000a0,0x608000000100)
freed by thread T0 here:
    #0 0x3ff8a34961b in free (/lib64/libasan.so.6+0xc961b)
    #1 0x3ff851d1651 in CRYPTO_free_ex_data crypto/ex_data.c:402
previously allocated by thread T0 here:
    #0 0x3ff8a34990d in malloc (/lib64/libasan.so.6+0xc990d)
    #1 0x3ff851d48c9 in CRYPTO_zalloc crypto/mem.c:191
SUMMARY: AddressSanitizer: heap-use-after-free crypto/provider_child.c:329 in ossl_provider_free_parent
Shadow bytes around the buggy address:
  0x100c0fffffffc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100c0fffffffd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100c0fffffffe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100c0ffffffff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100c1000000000: fa fa fa fa 00 00 00 00 00 00 00 00 00 00 00 00
=>0x100c1000000010: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd[fd]
  0x100c1000000020: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x100c1000000030: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x100c1000000040: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x100c1000000050: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x100c1000000060: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==443426==ABORTING
```

**Comment by mattcaswell** [comment]
Hmm. I cannot replicate it. How do you configure?

**Comment by juergenchrist** [comment]
> Hmm. I cannot replicate it. How do you configure?

Our CI uses this line:
```
./config --strict-warnings enable-asan enable-ubsan enable-zlib-dynamic enable-unit-test enable-md2 enable-rc5 enable-buildtest-c++ enable-weak-ssl-ciphers enable-ssl3 enable-ssl3-method -w
```
Not sure, but could this be s390x specific?  I doubt it, but I am not sure.

**Comment by juergenchrist** [comment]
Just tested on current master branch.  Problem is still there.

**Comment by mattcaswell** [comment]
With that config line I can replicate it. I think I understand the problem. PR coming up.

**Comment by mattcaswell** [comment]
Fix in #15286. Please try it out.

**Comment by juergenchrist** [comment]
> Please try it out.

Done.  I can confirm that this fixes the problem.  Thanks.
